### PR TITLE
feature: support math functions in FreeParameterExpression

### LINKED ISSUE #1252: Support math functions (`sin`, `cos`, `exp`, …) in `FreeParameterExpression`
This issue is part of **unitaryHACK26**. You have to be [registered](https://unitaryhack.dev/register/) to complete this issue.
Learn more about the PR submission process [here](https://unitaryhack.dev/hacker-guide/) and about unitaryHACK rules [here](https://unitary.design/rules/)!

**A note about AI Slop:** while we are open to collaboration with LLMs for unitaryHACK, fully AI-generated PRs are not acceptable. It is at the maintainers' discretion whether or not LLM-generated PRs are the right fit for the issues, and those that appear fully AI-generated may be immediately rejected. Read unitaryHACK's full AI Policy [here](https://unitaryhack.dev/ai-guide/) 

---

# Support math functions (`sin`, `cos`, `exp`, …) in `FreeParameterExpression`

## Summary

`FreeParameterExpression` supports basic arithmetic (`+`, `-`, `*`, `/`, `**`, unary `-`), but there is **no idiomatic way** to build expressions with math functions like `sin`, `cos`, `exp`, or `sqrt`. Users who force it with sympy get expressions that serialize to invalid OpenQASM 3 and fail on the device.

This issue tracks making symbolic math functions first-class: a helper API to construct them and a serializer that emits valid OpenQASM 3 for them.

## What works today

```python
from braket.parametric import FreeParameter
from braket.circuits import Circuit

x = FreeParameter("alpha")
Circuit().rx(0, 2 * x + 1).to_ir()    # basic arithmetic: ✅
```

### 1. No helper API — forced to import sympy

You cannot write:

```python
from braket.parametric import sin     # ImportError
```

The only way to build `sin(alpha)` today is:

```python
import sympy
from braket.parametric import FreeParameter, FreeParameterExpression

alpha = FreeParameter("alpha")
expr = FreeParameterExpression(sympy.sin(alpha.expression))
```

That requires a direct sympy dependency and knowledge of `.expression` as an internal attribute. Not something a user should have to reach for.

### 2. The string parser rejects function calls

Parallel to the `/` issue that #1251 fixes for `ast.Div`, the string parser has no case for `ast.Call`:

```python
>>> FreeParameterExpression("sin(alpha)")
ValueError: Unsupported string detected: <ast.Call object at ...>
```

### 3. Valid-looking constructions emit invalid OpenQASM 3

Even when users succeed in building an expression via sympy, several common functions serialize to strings that OpenQASM 3 rejects. Verified with `LocalSimulator`:

| Construction | Emitted in OpenQASM | OpenQASM 3 expects | Runs on `LocalSimulator`? |
| --- | --- | --- | --- |
| `sympy.sin(x)` | `sin(alpha)` | `sin(alpha)` | ✅ |
| `sympy.cos(x)` | `cos(alpha)` | `cos(alpha)` | ✅ |
| `sympy.exp(x)` | `exp(alpha)` | `exp(alpha)` | ✅ |
| `sympy.asin(x)` | `asin(alpha)` | `arcsin(alpha)` | ❌ `NameError: Subroutine asin is not defined` |
| `sympy.acos(x)` | `acos(alpha)` | `arccos(alpha)` | ❌ |
| `sympy.atan(x)` | `atan(alpha)` | `arctan(alpha)` | ❌ |
| `sympy.Mod(x, 2)` | `Mod(alpha, 2)` | `mod(alpha, 2)` | ❌ `Subroutine Mod is not defined` |
| `sympy.Abs(x)` | `Abs(alpha)` | (no OpenQASM 3 equivalent) | ❌ |
| `sympy.ceiling(x)` | `ceiling(alpha)` | `ceiling(alpha)` | ✅ |
| `sympy.floor(x)` | `floor(alpha)` | `floor(alpha)` | ✅ |

Root cause: `Gate._to_openqasm` formats gate parameters with `str(parameter)`, which goes through sympy's default printer. Sympy's canonical names for the inverse trig functions and `Mod`/`Abs` don't match OpenQASM 3's.

## Proposed solution

Both pieces must be implemented to get the bounty:

### (A) Module-level helper functions — *pure addition*

Export math helpers from `braket.parametric` for each [OpenQASM 3 built-in function](https://openqasm.com/language/types.html#built-in-functions):

```python
from braket.parametric import (
    FreeParameter,
    sin, cos, tan,
    arcsin, arccos, arctan,
    exp, log, sqrt,
    mod, ceiling, floor,
)

x = FreeParameter("alpha")
expr = sin(x / 2) ** 2 + cos(x / 2) ** 2
```

Each helper:

- Accepts `FreeParameterExpression | FreeParameter | Number`.
- Wraps the argument's `.expression` in the corresponding sympy function.
- Returns a `FreeParameterExpression`.

Implementation sketch:

```python
# src/braket/parametric/functions.py
import sympy
from numbers import Number
from braket.parametric.free_parameter_expression import FreeParameterExpression

def _wrap(sympy_fn, arg, *extra_args):
    inner = arg.expression if isinstance(arg, FreeParameterExpression) else arg
    return FreeParameterExpression(sympy_fn(inner, *extra_args))

def sin(x):     return _wrap(sympy.sin, x)
def cos(x):     return _wrap(sympy.cos, x)
# ...etc
def mod(x, m):  return _wrap(sympy.Mod, x, m)
```

### (B) OpenQASM-aware `str` / `repr` — the printer fix

Replace sympy's default string printer for `FreeParameterExpression.__str__` and `__repr__` with a `sympy.printing.StrPrinter` subclass that uses OpenQASM 3 names:

```python
class _OpenQASMPrinter(sympy.printing.StrPrinter):
    _FN_MAP = {
        sympy.sin:     "sin",
        sympy.cos:     "cos",
        sympy.tan:     "tan",
        sympy.asin:    "arcsin",
        sympy.acos:    "arccos",
        sympy.atan:    "arctan",
        sympy.exp:     "exp",
        sympy.log:     "log",
        sympy.sqrt:    "sqrt",
        sympy.Mod:     "mod",
        sympy.ceiling: "ceiling",
        sympy.floor:   "floor",
    }

    def _print_Function(self, expr):
        fn_type = type(expr)
        if fn_type not in self._FN_MAP:
            raise ValueError(f"No OpenQASM 3 equivalent for {fn_type.__name__}")
        name = self._FN_MAP[fn_type]
        args = ", ".join(self._print(a) for a in expr.args)
        return f"{name}({args})"
```

Because `Gate._to_openqasm` formats parameters via `str(parameter)`, this one change fixes the circuit-serialization path for every supported function. No change to `Gate._to_openqasm` or call sites needed.

`sympy.Abs`, `sympy.re`, `sympy.im`, and `sympy.conjugate` have no OpenQASM 3 equivalent — the printer should **raise** on those rather than silently emit invalid QASM. This is technically a behavior change, but today's output for those functions never actually runs on any device, so nothing was working to begin with.

## Minimal reproducer

```python
from braket.circuits import Circuit
from braket.circuits.serialization import IRType
from braket.parametric import FreeParameter, FreeParameterExpression
import sympy

alpha = FreeParameter("alpha")
expr = FreeParameterExpression(sympy.asin(alpha.expression))

circuit = Circuit().rx(0, expr).measure(0)
print(circuit.to_ir(ir_type=IRType.OPENQASM).source)
# OPENQASM 3.0;
# input float alpha;
# bit[1] b;
# qubit[1] q;
# rx(asin(alpha)) q[0];       # ← should be arcsin
# b[0] = measure q[0];

from braket.devices import LocalSimulator
LocalSimulator().run(circuit, inputs={"alpha": 0.5}, shots=10).result()
# NameError: Subroutine asin is not defined.
```

After this issue ships, the serialized line should read `rx(arcsin(alpha)) q[0];` and the circuit should run on `LocalSimulator`.

## Definition of done

- [ ] (A) `braket.parametric` exports `sin`, `cos`, `tan`, `arcsin`, `arccos`, `arctan`, `exp`, `log`, `sqrt`, `mod`, `ceiling`, `floor` as module-level functions returning `FreeParameterExpression`.
- [ ] (B) `str(FreeParameterExpression(sympy.asin(x)))` emits `arcsin(x)`; analogous fixes for `acos`/`atan`/`Mod`. Unsupported functions raise a clear error at stringify time.
- [ ] Unit tests for each new helper: constructor, round-trip, `subs` (partially and fully bound), OpenQASM emission.
- [ ] Integration test running `rx(sin(alpha/2)**2 + cos(alpha/2)**2)` on `LocalSimulator` with `alpha = π`.
- [ ] Docstring example using the new helpers.

## Not in scope

- Pulse-level serialization (`PulseSequence`, custom waveforms, `defcal`/`cal` blocks, calibration emission). `FreeParameterExpression._to_oqpy_expression` also crashes on `sympy.Function` subclasses, but fixing that requires separate `oqpy`/`openqasm3` AST lowering. If there's demand, it can be tracked in a follow-up issue.
- `sympy.Abs`, `sympy.re`, `sympy.im`, `sympy.conjugate`: no OpenQASM 3 equivalents. The helpers omit them; the printer raises on them.
- `popcount`, `rotl`, `rotr`: OpenQASM 3 has these for bit manipulation but they don't compose naturally with float gate angles. Can be added later if users ask for them.
- The `ast.Div` string-parser gap. Tracked and fixed in #1251.

## Prior work

- PR [#885](https://github.com/amazon-braket/amazon-braket-sdk-python/pull/885) — added `__truediv__` operator method.
- [qiskit-braket-provider#153](https://github.com/amazon-braket/qiskit-braket-provider/issues/153) — motivated the division fix in #885.

## Relevant code

- [`src/braket/parametric/free_parameter_expression.py`](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/src/braket/parametric/free_parameter_expression.py) — primary changes for (B).
- [`src/braket/parametric/__init__.py`](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/src/braket/parametric/__init__.py) — re-export helpers for (A).
- [`src/braket/parametric/free_parameter.py`](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/src/braket/parametric/free_parameter.py) — likely no changes.
- [`src/braket/circuits/gate.py`](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/src/braket/circuits/gate.py) `_to_openqasm` — verify (B) alone is enough; it formats parameters via `str(parameter)`.

## uH 2026

Please note that we're still waiting for Unitary Foundation to review and approve our issues, so these are subject to change.

### PR DESCRIPTION
**Pull request by border-b**
*Issue #, if available:*
Closes [#1252](https://github.com/amazon-braket/amazon-braket-sdk-python/issues/1252)

*Description of changes:*
Adds support for math functions in `FreeParameterExpression`.

- Adds `braket.parametric` helpers for `sin`, `cos`, `tan`, `arcsin`, `arccos`, `arctan`, `exp`, `log`, `sqrt`, `mod`, `ceiling`, and `floor`, returning `FreeParameterExpression` values without requiring direct SymPy use.
- Updates `str`/`repr` so supported SymPy functions serialize with OpenQASM names (`arcsin`, `arccos`, `arctan`, `mod`). Functions without an OpenQASM equivalent (`Abs`, `re`, `im`, `conjugate`) now raise `ValueError` instead of emitting invalid QASM.
  
*Testing done:*
`tox -e unit-tests`, `tox -e linters_check`, and `tox -e docs`.

**AI disclosure:** I used GPT-5.5 from OpenAI Codex as a support tool while working through the printer and helper tests. The implementation choices, final edits, and verification are mine. I reviewed the code and ran the tests listed above.

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md) doc
- [x] I used the PR title format described in [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#PR-title-format)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/README.md) and [API docs](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/braket/parametric/functions.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+) -> FreeParameterExpression:
+    sympy_args = [_expression(arg) for arg in args]
+    return FreeParameterExpression(function(*sympy_args))
+
+
+def sin(x: FreeParameterExpression | Number) -> FreeParameterExpression:
+    """Returns the sine of a free parameter expression.
+
+    Args:
+        x (FreeParameterExpression | FreeParameter | Number): The expression.
```
Comment (nilsquet) [inline src/braket/parametric/functions.py]: Annotations and docstrings disagree across all helpers — annotation is `FreeParameterExpression | Number` but docstrings list `FreeParameterExpression | FreeParameter | Number`. Since `FreeParameter` subclasses `FreeParameterExpression`, the annotation is already correct. Could you drop `FreeParameter` from the docstrings to match? That's also the existing style in `FreeParameterExpression.__init__`.

File: src/braket/parametric/free_parameter_expression.py [hunk 2]
```diff
@@ -17,11 +17,39 @@
...
+        sympy.exp: "exp",
+        sympy.log: "log",
+        sympy.Mod: "mod",
+        sympy.ceiling: "ceiling",
+        sympy.floor: "floor",
+    }
+
+    def _print_Function(self, expr: sympy.Function) -> str:
+        function_name = self._FN_MAP.get(expr.func)
+        if not function_name:
```
Comment (nilsquet) [inline src/braket/parametric/free_parameter_expression.py]: Nit: if not function_name: happens to work because all _FN_MAP values are non-empty strings, but `if function_name is None:` is more idiomatic for dict.get() results and a bit more defensive against future edits.

File: test/unit_tests/braket/devices/test_local_simulator.py [hunk 3]
```diff
@@ -689,6 +692,30 @@ def test_run_program_model_inputs():
     assert task.result() == GateModelQuantumTaskResult.from_object(GATE_MODEL_RESULT)
 
 
+def test_local_simulator_runs_sympy_inverse_trig_free_parameter_expression():
+    # Use StateVectorSimulator directly so OpenQASM validation runs.
```
Comment (nilsquet) [inline test/unit_tests/braket/devices/test_local_simulator.py]: Please remove this comment, because it's not adding value.

File: test/unit_tests/braket/devices/test_local_simulator.py [hunk 4]
```diff
@@ -689,6 +692,30 @@ def test_run_program_model_inputs():
...
+    circuit = Circuit().rx(0, expr).measure(0)
+
+    device = LocalSimulator(StateVectorSimulator())
+    result = device.run(circuit, inputs={"alpha": 0.5}, shots=10).result()
+
+    assert sum(result.measurement_counts.values()) == 10
+
+
+def test_local_simulator_runs_parametric_math_helper_expression():
+    # Use StateVectorSimulator directly so OpenQASM validation runs.
```
Comment (nilsquet) [inline test/unit_tests/braket/devices/test_local_simulator.py]: Please remove this comment, because it's not adding value.

File: src/braket/parametric/functions.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+"""Math helper functions for free parameter expressions.
+
+These helpers let you build symbolic OpenQASM math expressions without importing
+SymPy or touching the internal expression attribute.
+
+Examples:
+    >>> from braket.parametric import FreeParameter, cos, sin
```
Comment (nilsquet) [inline src/braket/parametric/functions.py]: The example builds an expression that isn't used then — readers can't tell what the helpers are good for from this. Could you replace it with something that shows the actual use case?

**Review by nilsquet — CHANGES_REQUESTED** [review]
Thanks for the contribution! The implementation is clean and the tests are thorough. Just one major request before merging: Please trim the test redundancy:
 
  There's significant overlap across the four test files — the helper names get checked at the printer level (`test_free_parameter_expression.py`), at the helper level (`test_functions.py`), and twice through `Circuit.to_ir` (`test_circuit.py`). Once the printer is unit-tested, going through `Circuit` doesn't exercise new logic — `Gate._to_openqasm` just calls `str(parameter)`.

**Comment by border-b** [comment]
Thanks for the review @nilsquet! I pushed an update addressing all the feedback. Trimmed the redundant tests and cleaned up the smaller nits. Let me know if you'd like any other changes.

File: test/unit_tests/braket/parametric/test_functions.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        (arctan, ()),
+        (exp, ()),
+        (log, ()),
+        (sqrt, ()),
+        (mod, (2,)),
+        (ceiling, ()),
+        (floor, ()),
+    ],
+)
+def test_helper_expression_round_trip(helper, extra_args):
```
Comment (nilsquet) [inline test/unit_tests/braket/parametric/test_functions.py]: This test can be removed. It asserts `FreeParameterExpression(expr.expression) == expr`, which only exercises the constructor and `__eq__` — same code path for all 12 parametrizations. If the constructor were broken, every other test in the file would fail first.

File: test/unit_tests/braket/parametric/test_functions.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+)
+def test_helper_constructs_expected_expression(helper, sympy_function, extra_args, expected):
+    alpha = FreeParameter("alpha")
+    expr = helper(alpha, *extra_args)
+    expected_expr = FreeParameterExpression(sympy_function(alpha.expression, *extra_args))
+
+    assert isinstance(expr, FreeParameterExpression)
+    assert expr == expected_expr
+    assert str(expr) == expected
+    assert repr(expr) == expected
```
Comment (nilsquet) [inline test/unit_tests/braket/parametric/test_functions.py]: This is already tested in `test_openqasm_function_names`. Please remove these lines.

File: test/unit_tests/braket/parametric/test_functions.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    expr = helper(alpha / 2, *extra_args)
+    rebuilt = FreeParameterExpression(expr.expression)
+
+    assert rebuilt == expr
+    assert str(rebuilt) == str(expr)
+
+
+def test_helpers_accept_numeric_inputs():
+    assert isinstance(sin(0), FreeParameterExpression)
+    assert isinstance(sqrt(4), FreeParameterExpression)
```
Comment (nilsquet) [inline test/unit_tests/braket/parametric/test_functions.py]: Remove. `sqrt`'s specialness is in the printer, not in numeric-input handling. `sin(0)` already covers unary, `mod(5, 2)` covers binary.

File: test/unit_tests/braket/parametric/test_functions.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+    assert isinstance(mod(5, 2), FreeParameterExpression)
+
+
+def test_sin_partial_and_full_substitution():
```
Comment (nilsquet) [inline test/unit_tests/braket/parametric/test_functions.py]: This test can be removed — `test_unary_helper_partial_and_full_substitution` already covers `sin` as its first parametrized case. The remaining string assertion (`str(partial) == "sin(beta)"`) doesn't add anything: the parametrized test's `partial == helper(beta)` is structural equality, which already
  implies the two print the same.

File: test/unit_tests/braket/parametric/test_functions.py [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+    floor,
+    log,
+    mod,
+    sin,
+    sqrt,
+    tan,
+)
+
+
+@pytest.mark.parametrize(
```
Comment (nilsquet) [inline test/unit_tests/braket/parametric/test_functions.py]: The helper⇄SymPy-function mapping is duplicated across `test_helper_constructs_expected_expression` and `test_unary_helper_partial_and_full_substitution`. Could you pull it into a module-level constant so the two lists can't drift if a helper is added or renamed?

File: test/unit_tests/braket/parametric/test_functions.py [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
```
Comment (nilsquet) [inline test/unit_tests/braket/parametric/test_functions.py:13]: Could you add a test for nested helper composition, e.g. `sin(cos(alpha))` and `arcsin(sqrt(alpha))`? The printer's `_print_Function` recurses via `self._print(arg)`, but no current test exercises that path — a regression in the recursion would slip through.

Also, could you add a test for multi-function expressions with mixed arithmetic, e.g. `sin(alpha) + cos(beta) * sqrt(gamma)`? It verifies the printer composes correctly when Function and Pow nodes are interleaved with `Add`/`Mul`. SymPy may reorder commutative operands, so check substring containment rather than exact string match.

File: test/unit_tests/braket/devices/test_local_simulator.py [hunk 12]
```diff
@@ -689,6 +692,28 @@ def test_run_program_model_inputs():
     assert task.result() == GateModelQuantumTaskResult.from_object(GATE_MODEL_RESULT)
 
 
+def test_local_simulator_runs_sympy_inverse_trig_free_parameter_expression():
```
Comment (nilsquet) [inline test/unit_tests/braket/devices/test_local_simulator.py:695]: The two `test_local_simulator.py` tests verify the OpenQASM is *parseable* but don't pin the exact source format. Could you add one assertion that `circuit.to_ir().source` contains e.g. `arcsin(alpha)`? Catches regressions where the printer produces something parseable but not the intended OpenQASM name.

**Review by nilsquet — CHANGES_REQUESTED** [review]
Thanks for the changes, this goes into the right direction. I added some concrete examples how we can clean-up the tests even further. Please have a look.

File: test/unit_tests/braket/parametric/test_functions.py [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+def test_sqrt_constructs_expected_expression():
+    # SymPy represents sqrt as a Pow expression, so it is verified separately.
+    alpha = FreeParameter("alpha")
+    expr = sqrt(alpha)
+
+    assert isinstance(expr, FreeParameterExpression)
+    assert expr == FreeParameterExpression(sympy.sqrt(alpha.expression))
+    assert str(expr) == "sqrt(alpha)"
+    assert repr(expr) == "sqrt(alpha)"
```
Comment (border-b) [inline test/unit_tests/braket/parametric/test_functions.py]: Removing these too, as this is tested in `test_openqasm_sqrt_function_name`

**Comment by border-b** [comment]
Thanks @nilsquet! Pushed an update addressing the latest feedback:

- Trimmed the redundant tests and pulled the helper -> SymPy pairing into a single `HELPER_TO_SYMPY` constant.
- Added the missing coverage for printer recursion tests plus a `to_ir(...).source` assertion pinning `arcsin(alpha)`.

I also re-checked the rest of the new/changed tests with the same lens and removed one more duplicate (`test_sqrt_constructs_expected_expression`, already covered by `test_openqasm_sqrt_function_name`, flagged inline). Nothing else stood out.

Happy to keep iterating if there's anything you'd like done differently.

**Review by nilsquet — APPROVED** [review]
Thanks for the thorough iteration, @border-b! The trimmed tests, the centralized `HELPER_TO_SYMPY` mapping, and the added composition / `to_ir().source` coverage land this in a good place. Approving.
